### ISSUE TITLE
Support connect to wss with custom domain

### ISSUE THREAD

**Issue by WiSaGaN** [cyderize/rust-websocket#167]
Sometimes I need to do async connect to a wss with ip address, but need to provide the domain name for secure connection purpose, but we only have `pub fn async_connect(self, ssl_config: option<TlsConnector>, handle: &Handle)` now. Would be great if we can provide an optional domain name or a new methodd "async_connect_with_domain.

If that is ok, I can provide a pull request.

**Comment by vi** [comment]
Maybe you can use `async_connect_on`, with a pre-connected SSL socket?

**Comment by git-blame** [comment]
Does `async_connect_on` work with an existing stream (TLS or not)? Due to issue #164, I tried to switch to `async_connect_on` from `async_connect`.

As a first step, I tried to connect with a `websocket::client::async::TcpStream` and then pass it to `async_connect_on`. But outbound messages only work occasionally and inbound messages are never seen. Note that everything looks ok at the network layer via a network sniffer. That is, inbound messages from server show up on tcpdump.

**Comment by WiSaGaN** [comment]
Actually I think using a pre constructed TLS stream does not work, as the domain seems still needed in websocket handshake. It will result a response error of "status code not switching protocols" error. Closing this for now.

**Comment by vi** [comment]
With pre-supplied TLS stream you also need to supply the full URI. "Not switching protocols" should also be investigated on the server.

There are total three domains involved:

1. Domain name to be resolved to IP;
2. Domain name for SNI for SSL connection;
3. Domain name for `Host:` in websocket's Upgrade query.

**Comment by vi** [comment]
> But outbound messages only work occasionally and inbound messages are never seen.

Maybe it's some bug? Once it connected, I assume it should not matter anymore if it was `async_connect` or `async_connect_on`... Maybe thread is just getting blocked somewhere?

**Comment by WiSaGaN** [comment]
@vi, thanks for the explanation, my original plan was to only use domain name for SSL, but use ip in TCP connection, and websocket, as my goal is to specify IP directly instead of relying on DNS dynamic routing. At last, I just switched to use IP for TCP connection, and use the domain for the rest, then it works, which actually also better documents the intention. I guess the not switching protocol was caused by step 3 in your list where the server expected a domain name in the query rather than the IP.

**Comment by git-blame** [comment]
Hmm, I'll have to try `async_connect_on` again. As a first step, I simply tried to modify _examples/async-client.rs_ so it shouldn't have any threading issue. I mean, there is a thread that waits for user input. But the runner is in the main thread. @vi, did you do anything special for your stream. For example, I just noticed in function `async_connect`, web-socket passes a `Boxed` stream to `async_connect_on`? I didn't in my previous attempt.
